### ISSUE TITLE
readline() looses data red so far if timeout happens

### ISSUE THREAD

**Issue by haizaar** [ronf/asyncssh#140]
Good day,

asyncssh seem to loose data buffered so far when reading a line if timeout happens. Consider the following:

```
cmd = """
echo fast line
echo -n 'slow line'; sleep 1
echo; echo "fast line"
"""
async def run_client():
    async with asyncssh.connect("localhost") as conn:
        async with conn.create_process(cmd, term_type="xterm-color") as process:
            while process.exit_status is None:
                try:
                    result = await asyncio.wait_for(process.stdout.readline(), timeout=.5)
                    print(result, end="")
                except asyncio.TimeoutError as exc:
                    print("Got timeout - retrying")
            print("Remote process finished with exit status {}".format(process.exit_status))
```

The output is as follows:

```
fast line
Got timeout - retrying

fast line
Remote process finished with exit status 0
```

I.e the slow line got lost. This become a real problem with running remote third-party scripts that switch to interactive mode when an error occurs, e.g.

```
cmd = """
echo "So fall all good!"
echo -n 'Oops! Error occured due to matian invasion inbound. Press enter to retry'; read
"""
```

In this case there is no way for me to log the error, since it appears on the already lost line. 

From a quick look at stream.py:SSHStreamSession.readuntil(), it looks like it actually consumes the red buffer as it progresses with its search for separator. so if it fails to return the consumed data (due to timeout), this data eventually gets lost. May be it will be better if readuntil() would not actually consume self._read_buf, until the whole line is present in it. What do you think?

**Comment by ronf** [comment]
Thank you for the report - I see the issue, and agree with you that the problem is with readuntil() consuming the data even before it is ready to return. Changing it to keep the input in recv_buf until it finds the delimiter seems like a reasonable approach. A consequence of that would be that it would be impossible to read more with readuntil() than whatever the SSH channel's window size is set to, but that's probably a reasonable trade-off. 

A workaround for now would be to use a plain read() instead of readline() or readuntil() in cases where you are possibly expecting to receive a prompt. While you might need to call read multiple times in this case, you wouldn't need a timeout if you were willing to wait for either more data or for the remote end to send an EOF, and you wouldn't be relying on the timeout to telly you that you were at the end of a "prompt" (which isn't really reliable). That said, discarding data on a timeout is clearly wrong, so I'll work on fixing that.

**Comment by ronf** [comment]
I just committed a potential fix (8f47a09) for this issue in the "develop" branch. I still need to go back and handle the case where the receive window fills up with no delimiter being sent, but with this new implementation, the data loss should no longer occur. If you are able to test it, I'd love to hear if this works for you. It looked good in the unit test I wrote.

**Comment by haizaar** [comment]
Thanks for such a quick fix! I've tested it with my "martian" command, and now it prints the prompt line as well:

Before:
```
So fall all good!
Got timeout - killing process
Last lines, if any: ('', '')
Remote process finished with exit status -1
```

After:
```
So fall all good!
Got timeout - killing process
Oops! Error occured due to matian invasion inbound. Press enter to retryLast lines, if any: ('', '')
Remote process finished with exit status -1
```

That `Oops! Error occured due to matian invasion inbound. Press enter to retry` is what was previously lost.

Thanks again. Looking forward to have it in master. 

My SW is in development, so the workaround you suggested is not that urgent for me (it also means I would implement my own version of readline(), more or less :). In my scenario, I expect a remote command to finish within a certain timeout, and if it does not I just want to log whatever it had a chance to print and terminate it, i.e. I don't care about the prompt specifically, but it causes the problem for me.

As for the SSH channel's window size limitation you mentioned - I also agree that it's reasonable. After all, whoever tries to read lines, probably does not expect them to be over 2mb long.

**Comment by ronf** [comment]
Thanks for the confirmation that it worked. I'll definitely roll this into my next patch release.

Some other thoughts on this:

If you're willing to wait until the process exits on its own in all cases and just want to collect up all the output it generated, you might just want to call conn.run(cmd, term_type='xterm-color') here and not worry about having your own read loop at all. This wouldn't let you display the output incrementally as it was produced, but it would return all of the output of both stdout and stderr in the SSHCompletedProcess object returned by the run() call when the process exited without you having to read it or have any timeouts, and it wouldn't care at all about if there was output without a newline at the end. Actually, if the remote end is closing the connection, even readline() should have returned as soon as the close happened, though, without needing a timeout.

Alternately, if you don't need to process the output a line at a time, you could change your example here to simply call something like read(1024) (or any other positive value) instead of readline() and then it would return immediately each time ouput was available,  with or without a newline.

Finally, if you wanted to enforce your own overall timeout on a command and not wait for the remote system to close the session, you might want to look into the collect_output() method on SSHProcess. It will return whatever output is waiting in the receive buffer without blocking. So, you could do something like wrap an asyncio.wait_for() around process.wait() and use collect_output() to pick up all the output generated so far when a timeout occurred. You could then call close().

**Comment by haizaar** [comment]
Thanks for the hints.
I prefer to stream the output, because I want to watch it as it happens. I'm running remote installers, that take time to complete, so I need something to look at while it happens :)

I'm currently doing is what you are actually suggesting in your last paragraph, only with readline() thrown in:

```python
async with self.connection.create_process(cmd, term_type="xterm-color") as process:
    while process.exit_status is None:
        try:
            start = time.monotonic()
            line = await asyncio.wait_for(process.stdout.readline(), timeout=timeout)
            logger.info(line)

            if timeout is not None:
                timeout -= (time.monotonic() - start)
                if timeout <= 0:
                    raise asyncio.TimeoutError()

        except asyncio.TimeoutError:
            logger.warn("Timeout on command execution")
            logger.warn("Closing stdin and collecting last output if any")
            process.stdin.close()
            process.kill()
            raise
        finally:
            for line in process.collect_output()[0].splitlines():
                logger.info(line)
```

**Comment by ronf** [comment]
Yeah, I figured that's why you might want to read as you went along, and if you're passing read data to the logger I can see why you want it to be a line at a time.

One caveat here: If you are using OpenSSH as your server, "process.kill()" doesn't work on that, as OpenSSH doesn't implement the SSH "signal" message (see https://bugzilla.mindrot.org/show_bug.cgi?id=1424). It may not matter much since you're closing the channel, though -- at that point, OpenSSH could potentially figure out on its own that it should kill the child process.

Also, if you do call kill(), you really shouldn't call it after you call close() on stdin. The stdin close turns into a channel close, and the process kill turns into a channel kill, but calling any method on the channel after a close will generally either be ignored or explicitly return an error. In this case, it happens to ignore the call, but that means that even if the server supported it, the kill request would not be sent. You'd need to move it before the close() for that to work.

**Comment by haizaar** [comment]
Oh, thanks for that! Indeed there is an OpenSSH server on the other end, so I guess I'll remove the kill() invocation all together.

I also would like to express my great appreciation for developing this library. I think you really nailed it with the API (which I find to be one the hardests part in our trade) - it's really easy and succinct for the common use cases.

**Comment by ronf** [comment]
Thanks so much!

It's almost hard to imagine at this point, but this actually started out as an asyncore-based project about 5 years ago, and at that point I actually hadn't even heard of asyncio! Everything began as pure callbacks, with no coroutines at all. I converted it over to asyncio before I made it public, but the early versions of it didn't even have a high-level "stream" API, requiring you to provide custom SSHClient/SSHServer and session classes which implemented all the callbacks. Little by little, though, I've managed to build up higher-level APIs that are easier to use.

! It took me a couple of iterations on a few of the API details and I'm still carrying a bit of legacy baggage as a result of that in terms of all the different types of factory objects I have to support, but I'm pretty happy with how SSHProcess turned out. My goal was definitely to keep the code as dead simple as possible for the "easy" stuff, while still providing knobs for the more complex cases. It's nice to hear that you see it that way!

BTW, I checked in a small change tonight to handle the case of maxing out the receive window without finding a delimiter in a readuntil() call. That now raises IncompleteReadError, just like it does when it receives EOF or something like a break, signal, or terminal size change, returning the full window's worth of partial data in the exception object. That data is removed from the buffer, so you can then call readuntil() again to keep reading if you want to, and then append the pieces together. This works for readline() as well, but that simply returns the partial line (without a trailing '\n') rather than raising an exception, to be consistent with the way readline() works in these other cases (and the way it works on asyncio.StreamReader).

**Comment by haizaar** [comment]
(Does it happen to you that you write a message and forget to send it? - I actually wrote this yesterday)

I totally feel you regarding ongoing rewrites - went through kind of similar process with [mokoko](https://github.com/FSX/momoko) project.

As for your explanation about IncompleteReadError - sounds totally reasonable to me.

**Comment by ronf** [comment]
Yeah, that does look similar to what happened here.

Thanks for the feedback, and for your initial bug report!

These changes will go in the next release, and I'll update this issue as soon as that's available. There's one other bit of work I want to finish before then, but I should be able to get something out in the next few weeks - maybe sooner if things go well.

**Comment by ronf** [comment]
The bug fixes discussed here are now all available in AsyncSSH 1.13.0.
